clippy_lints/src/ptr_offset_with_cast.rs

```diff
@@ -0,0 +1,130 @@
+    pub PTR_OFFSET_WITH_CAST,
+    style,
+    "uneeded pointer offset cast"
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct Pass;
+
+impl lint::LintPass for Pass {
+    fn get_lints(&self) -> lint::LintArray {
+        lint_array!(PTR_OFFSET_WITH_CAST)
+    }
+}
+
+impl<'a, 'tcx> lint::LateLintPass<'a, 'tcx> for Pass {
+    fn check_expr(&mut self, cx: &lint::LateContext<'a, 'tcx>, expr: &'tcx hir::Expr) {
+        // Check if the expressions is a ptr.offset method call
+        let [receiver_expr, arg_expr] = match expr_as_ptr_offset_call(cx, expr) {
+            Some(call_arg) => call_arg,
+            None => return,
+        };
+
+        // Check if the parameter to ptr.offset is a cast from usize to isize
+        let cast_lhs_expr = match expr_as_cast_from_usize(cx, arg_expr) {
+            Some(cast_lhs_expr) => cast_lhs_expr,
+            None => return,
+        };
+
+        utils::span_lint_and_sugg(
+            cx,
+            PTR_OFFSET_WITH_CAST,
+            expr.span,
+            "use of `offset` with a `usize` casted to an `isize`",
+            "try",
+            build_suggestion(cx, receiver_expr, cast_lhs_expr),
+        );
+    }
+}
+
+// If the given expression is a cast from a usize, return the lhs of the cast
+fn expr_as_cast_from_usize<'a, 'tcx>(
+    cx: &lint::LateContext<'a, 'tcx>,
+    expr: &'tcx hir::Expr,
+) -> Option<&'tcx hir::Expr> {
+    if let hir::ExprKind::Cast(ref cast_lhs_expr, _) = expr.node {
+        if is_expr_ty_usize(cx, &cast_lhs_expr) {
+            return Some(cast_lhs_expr);
+        }
+    }
+    None
+}
+
+// If the given expression is a ptr::offset method call, return the receiver and the arg of the
+// method call.
+fn expr_as_ptr_offset_call<'a, 'tcx>(
+    cx: &lint::LateContext<'a, 'tcx>,
+    expr: &'tcx hir::Expr,
+) -> Option<[&'tcx hir::Expr; 2]> {
+    if let hir::ExprKind::MethodCall(ref path_segment, _, ref args) = expr.node {
+        if path_segment.ident.name == "offset" && is_expr_ty_raw_ptr(cx, &args[0]) {
+            return Some([&args[0], &args[1]]);
+        }
+    }
+    None
+}
+
+// Is the type of the expression a usize?
+fn is_expr_ty_usize<'a, 'tcx>(
+    cx: &lint::LateContext<'a, 'tcx>,
+    expr: &hir::Expr,
+) -> bool {
+    cx.tables.expr_ty(expr).sty == ty::TyKind::Uint(ast::UintTy::Usize)
+}
+
+// Is the type of the expression a raw pointer?
+fn is_expr_ty_raw_ptr<'a, 'tcx>(
+    cx: &lint::LateContext<'a, 'tcx>,
+    expr: &hir::Expr,
+) -> bool {
+    if let ty::RawPtr(..) = cx.tables.expr_ty(expr).sty {
+        true
+    } else {
+        false
+    }
+}
+
+fn build_suggestion<'a, 'tcx>(
+    cx: &lint::LateContext<'a, 'tcx>,
+    receiver_expr: &hir::Expr,
+    cast_lhs_expr: &hir::Expr,
+) -> String {
+    match (
+        utils::snippet_opt(cx, receiver_expr.span),
+        utils::snippet_opt(cx, cast_lhs_expr.span)
+    ) {
+        (Some(receiver), Some(cast_lhs)) => format!("{}.add({})", receiver, cast_lhs),
+        _ => String::new(),
```

Comment:
if the snippet is not obtainable, instead of reporting an empty snippet, we should not produce any suggestion at all

Reply:
https://github.com/rust-lang-nursery/rust-clippy/pull/3104/commits/61c20c148e926310f011895693f88cafb5a26539